go/arrow/array/util_test.go

```diff
@@ -0,0 +1,406 @@
+	}, nil), strings.NewReader(jsonStr))
+	assert.NoError(t, err)
+	defer rec.Release()
+
+	assert.EqualValues(t, 3, rec.NumCols())
+	assert.EqualValues(t, 3, rec.NumRows())
+
+	data, err := json.Marshal(rec)
+	assert.NoError(t, err)
+	assert.JSONEq(t, jsonStr, string(data))
+}
+
+func TestTimestampsJSON(t *testing.T) {
+	tests := []struct {
+		unit    arrow.TimeUnit
+		jsonstr string
+		values  []arrow.Timestamp
+	}{
+		{arrow.Second, `["1970-01-01", "2000-02-29", "3989-07-14", "1900-02-28"]`, []arrow.Timestamp{0, 951782400, 63730281600, -2203977600}},
+		{arrow.Nanosecond, `["1970-01-01", "2000-02-29", "1900-02-28"]`, []arrow.Timestamp{0, 951782400000000000, -2203977600000000000}},
+	}
+
+	for _, tt := range tests {
+		dtype := &arrow.TimestampType{Unit: tt.unit}
+		bldr := array.NewTimestampBuilder(memory.DefaultAllocator, dtype)
+		defer bldr.Release()
+
+		bldr.AppendValues(tt.values, nil)
+		expected := bldr.NewArray()
+		defer expected.Release()
+
+		arr, _, err := array.FromJSON(memory.DefaultAllocator, dtype, strings.NewReader(tt.jsonstr))
+		assert.NoError(t, err)
+		defer arr.Release()
+
+		assert.Truef(t, array.ArrayEqual(expected, arr), "expected: %s\ngot: %s\n", expected, arr)
+	}
+}
+
+func TestDateJSON(t *testing.T) {
+	t.Run("date32", func(t *testing.T) {
+		bldr := array.NewDate32Builder(memory.DefaultAllocator)
+		defer bldr.Release()
+
+		jsonstr := `["1970-01-06", null, "1970-02-12"]`
+
+		bldr.AppendValues([]arrow.Date32{5, 0, 42}, []bool{true, false, true})
+		expected := bldr.NewArray()
+		defer expected.Release()
+
+		arr, _, err := array.FromJSON(memory.DefaultAllocator, arrow.FixedWidthTypes.Date32, strings.NewReader(jsonstr))
+		assert.NoError(t, err)
+		defer arr.Release()
+
+		assert.Truef(t, array.ArrayEqual(expected, arr), "expected: %s\ngot: %s\n", expected, arr)
+
+		data, err := json.Marshal(arr)
+		assert.NoError(t, err)
+		assert.JSONEq(t, jsonstr, string(data))
+	})
+	t.Run("date64", func(t *testing.T) {
+		bldr := array.NewDate64Builder(memory.DefaultAllocator)
+		defer bldr.Release()
+
+		jsonstr := `["1970-01-02", null, "2286-11-20"]`
+
+		bldr.AppendValues([]arrow.Date64{86400000, 0, 9999936000000}, []bool{true, false, true})
+		expected := bldr.NewArray()
+		defer expected.Release()
+
+		arr, _, err := array.FromJSON(memory.DefaultAllocator, arrow.FixedWidthTypes.Date64, strings.NewReader(jsonstr))
+		assert.NoError(t, err)
+		defer arr.Release()
+
+		assert.Truef(t, array.ArrayEqual(expected, arr), "expected: %s\ngot: %s\n", expected, arr)
+
+		data, err := json.Marshal(arr)
+		assert.NoError(t, err)
+		assert.JSONEq(t, jsonstr, string(data))
+	})
+}
+
+func TestTimeJSON(t *testing.T) {
+	tententen := 60*(60*(10)+10) + 10
+	tests := []struct {
+		dt       arrow.DataType
+		jsonstr  string
+		valueadd int
+	}{
+		{arrow.FixedWidthTypes.Time32s, `[null, "10:10:10"]`, 123},
+		{arrow.FixedWidthTypes.Time32ms, `[null, "10:10:10.123"]`, 456},
```

Comment:
oh is the formatter smart to not extend 0 values on the decimal?

Reply:
Right, that's the difference between using 0's or 9's in the format string. Using 0's will extend it to the exact precision stated, using 9's will truncate trailing zeros.